src/hotspot/share/ci/ciArrayKlass.cpp

```diff
@@ -104,10 +105,36 @@ bool ciArrayKlass::is_leaf_type() {
 ciArrayKlass* ciArrayKlass::make(ciType* element_type) {
   if (element_type->is_primitive_type()) {
     return ciTypeArrayKlass::make(element_type->basic_type());
-  } else if (element_type->flatten_array()) {
-    return ciFlatArrayKlass::make(element_type->as_klass());
   } else {
-    return ciObjArrayKlass::make(element_type->as_klass());
+    return make(element_type->as_klass(), element_type->is_null_free());
+  }
+}
+
+ciArrayKlass* ciArrayKlass::make(ciKlass* klass, bool null_free) {
+  if (null_free) {
+    if (klass->is_loaded()) {
+      bool is_array_flattened = false;
+      GUARDED_VM_ENTRY(
+        EXCEPTION_CONTEXT;
+        Klass* ak = InlineKlass::cast(klass->get_Klass())->null_free_inline_array_klass(THREAD);
+        if (HAS_PENDING_EXCEPTION) {
+          CLEAR_PENDING_EXCEPTION;
+        } else {
+          if (ak != NULL && ak->is_flatArray_klass()) {
+            is_array_flattened = true;
+          }
+        }
+      )
+      if (is_array_flattened) {
+        return ciFlatArrayKlass::make(klass);
+      } else {
+        return ciObjArrayKlass::make(klass, true);
+      }
+    } else {
+      return ciEnv::unloaded_ciobjarrayklass();
```

Comment:
Can't we delegate handling the unloaded case to `ciObjArrayKlass::make(klass)` below? I.e. check for `if (null_free && klass->is_loaded())` above.

Reply:
I can change that. This was the way it was implemented initially, then I changed it to have the whole logic in a single method, but I don't have a strong opinion about this one.
Updated in the latest commit.